include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   */
+  std::map<CellIteratorType, CellVector<DataType> > map;
+
+};
+
+
+/**
+ * An abstract class which specifies requirements for data on
+ * a single quadrature point to be transferable during refinement.
+ * One possible implementation is via the restriction/prolongation matrices
+ * as implemented in ContinuousQuadratureDataTransfer class.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+class TransferableQuadraturePointData
+{
+public:
+  /**
+   * Default constructor.
+   */
+  TransferableQuadraturePointData();
+
+  /**
+   * Default virtual destructor.
+   */
+  virtual ~TransferableQuadraturePointData();
+
+  /**
+   * Returns the total number of values which will be
+   * packed/unpacked from the user's DataType class. Consequently it is also
+   * the size of the vectors in pack_values() and unpack_values() .
+   */
+  virtual unsigned int number_of_values() const = 0;
+
+  /**
+   * Pack all data stored in this class into a vector @p values .
+   * This vector will contain all scalar and/or Tensorial data local to each
+   * quadrature point.
+   *
+   * @p values will be resized to number_of_values() .
+   */
+  virtual void pack_values(std::vector<double> &values) const = 0;
+
+  /**
+   * The opposite of the above, namely
+   * unpack a vector @p values into the data stored in this class.
+   *
+   * @p values will be resized to number_of_values() .
+   */
+  virtual void unpack_values(const std::vector<double> &values) = 0;
+};
+
+
+#ifdef DEAL_II_WITH_P4EST
+namespace distributed
+{
+  /**
+   * A class for the transfer of continuous data stored at quadrature points when
+   * performing h-adaptive refinement of parallel::distributed::Triangulation .
+   *
+   * <h3>Implementation details</h3>
+   *
+   * This class implements the transfer of the quadrature point data between
+   * cells in case of adaptive refinement. That also includes automatic shipping of
+   * information between different processors.
+   *
+   * To that end, the constructor of the class is provided with three main objects:
+   * scalar FiniteElement @projection_fe, @p mass_quadrature and @p data_quadrature
+   * Quadrature rules.
+   * First, the data located at @p data_quadrature of each cell is L2-projected
+   * to the continuous space defined by a single FiniteElement @projection_fe .
+   * In  doing so the mass matrix of this element is required, which will be calculated
+   * with the @p mass_quadrature rule . Should the cell now belong to another processor,
+   * the data is then sent to this processor. The class makes use of a feature of p4est
+   * (and parallel::distributed::Triangulation) that allows one to attach information
+   * to cells during mesh refinement and rebalancing.
+   * On receiving information on the target cell, the data is projected back to the
+   * quadrature points. In the case that local refinement is performed, we have to first
+   * project local DoF values of the parent element to each child.
```

Comment:
The description in this paragraph suggests that "we have to" do it this way. But that's not quite true. This class just *chooses to implement* things this way. Maybe replace sentences such as "we have to first project..." by "this class first projects..."